build-scripts/use-mirror/set-v4-client-latest.sh

```diff
@@ -0,0 +1,107 @@
+MAJOR_NEXT_MINOR=$(echo ${MAJOR_MINOR} |awk -F '[.-]' '{print $1 "." $2+1}')  # e.g. 4.3 -> 4.4
+
+
+function create_latest_links {
+    local client_base_dir="$1" # e.g. /srv/pub/openshift-v4/s390x/clients/ocp-dev-preview
+
+    # Does this client base directory even exist? (it may not if this arch hasn't built clients yet)
+    if [[ ! -d "${client_base_dir}" ]]; then
+        echo "Unable to find client base directory: ${client_base_dir} ; will not update latest"
+        return 1
+    fi
+
+    cd "${client_base_dir}"
+
+    # If the release does not exist in the directory
+    if [[ ! -d "${RELEASE}" ]]; then
+        echo "Unable to find ${RELEASE} directory in ${PWD} ; will not update latest"
+        return 1
+    fi
+
+    MAJOR_MINOR_LATEST="latest-${MAJOR_MINOR}"
+
+    # Here's the easy part. The caller told us what is latest for a given major.minor; set it.
+    ln -svfn ${RELEASE} ${MAJOR_MINOR_LATEST}
+    echo "${MAJOR_MINOR_LATEST}  now points to ${RELEASE} in ${client_base_dir}"
+
+    # Here's the harder part - is this major.minor the latest Y stream? If it is, we need to set
+    # the overall 'latest'. We already calculated MAJOR_NEXT_MINOR  (e.g. "4.5") so see if that
+    # exists someone on the mirror.
+    if ls -d "${MAJOR_NEXT_MINOR}".*/ > /dev/null  2>&1; then
+        echo "This is not the highest Y release -- will not set overall latest"
+        return 0
+    fi
+
+    ln -svfn ${RELEASE} latest
+    echo "Overall latest now points to ${RELEASE} in ${client_base_dir}"
+    return 0
+}
+
+
+# We are transitioning to multi-arch builds. At the time of this writing, directories like
+# ${BASE_DIR}/${arch}/clients don't exist -- only ${BASE_DIR}/clients does. This script attempts to account
+# for the present and future. if x86_64 doesn't exist yet, we assume clients/ contains the x864_64 payload.
+
+if [[ ! -d "${BASE_DIR}/x86_64" ]]; then
+    echo "Can't find x86_64 arch; assuming single arch mode is still in place"
+    create_latest_links ${BASE_DIR}/clients/${CLIENT_TYPE}
+else
+
+    # Otherwise, multi-arch is enabled. Run the links logic for each architecture passed in.
+    cd "${BASE_DIR}"
+    MODE="${ARCHES}"
+    if [[ "${MODE}" == "all" || "${MODE}" == "any" ]]; then
+        # Find any subdirectory with a clients directory; it should be an arch
+        ARCHES=$(dirname $(ls -d */clients/))
+    fi
+
+    for arch in ${ARCHES}; do
+        target_dir="${BASE_DIR}/${arch}/clients/${CLIENT_TYPE}"
+        # Check that the clients for this arch exists.
+        if [[ ! -d "${target_dir}" ]]; then
+            echo "Unable to find clients directory under: ${target_dir}"
+            continue
+        fi
+
+        if ! create_latest_links "${target_dir}"; then
+            echo "Failure creating links in ${target_dir}"
+            if [[ "${MODE}" != "any" ]]; then
+                echo "This was a required operation; exiting with failure"
+                exit 1
```

Comment:
Should `/usr/local/bin/push.pub.sh` be called if this happens? Considering some symlinks might be already changed on use-mirror, those change could be pushed in succeeding runs.

Reply:
That's a good point and I had not considered it. But I think I would prefer to give a human a chance to try to figure out what went wrong before pushing that unexpected state to the world. It still might, if something else does a push, but it at least gives us a window to investigate.

Reply:
Makes sense. Thanks.